nearcore/src/config.rs

```diff
@@ -411,15 +420,23 @@ impl Config {
             );
         }
 
-        assert!(
-            config.archive || config.save_trie_changes,
-            "Configuration with archive = false and save_trie_changes = false is not supported \
-            because non-archival nodes must save trie changes in order to do do garbage collection."
-        );
-
+        config.validate()?;
         Ok(config)
     }
 
+    /// Does semantic config validation.
+    /// This is the place to check that all config values make sense and fit well together.
+    /// `validate()` is called every time `config.json` is read.
+    fn validate(&self) -> Result<(), ConfigValidationError> {
+        if self.archive == self.save_trie_changes {
+            Err(ConfigValidationError::TrieChanges)
+        } else {
+            Ok(())
+        }
+        // TODO: Add more config validation.
+        // TODO: Validate `ClientConfig` instead.
```

Comment:
+1 for validating `ClientConfig` instead. Ideally, the validation should happen as soon as ClientConfig is loaded.

Reply:
Do you mind if I leave this to a follow-up code change?